### ISSUE TITLE
Any chance of ever having nice, clean, declarative steps again?

### ISSUE THREAD

**Issue by benko** [openshift/jenkins-client-plugin#227]
So, great. The old plugin is obsolete now. The only way to stay current is to replace every single step from our old pipelines with 6-15 lines of curly brace and flow control noise.

Because this:

        node('maven') {
          stage('build') {
            openshiftBuild(buildConfig: 'simple', showBuildLogs: 'true')
          }
          stage('deploy') {
            openshiftDeploy(deploymentConfig: 'simple')
          }
        }

now looks like this:

        node('maven') {
          stage('start a build') {
            openshift.withCluster() {
              openshift.withProject() {
                openshift.selector("bc/simple").startBuild()
                def x = openshift.selector("bc/simple").related("builds")
                x.untilEach(1) {
                  return (it.object().status.phase == "Complete")
                }
              }
            }
          }
          stage('start a deployment') {
            openshift.withCluster() {
              openshift.withProject() {
                openshift.selector("dc/simple").rollout().latest()
                def x = openshift.selector("dc/simple").related("pods")
                x.untilEach(1) {
                  return (it.object().status.phase == "Running")
                }
              }
            }
          }
        }

And mind, this is scripted, not declarative. Declarative just adds another 10 lines of curly brace noise per stage. And it's incomplete in status checks, so I don't really know what's going to happen when the build fails. I don't want to program a timeout *and* check for all possible exit codes in untilEach(). I want to start a build, and I want the next step to happen if the build succeeds. And I want the rest of the pipeline to blow up if it doesn't, and have a couple of post steps where I handle it.

And by saying "I", I mean hundreds of users that have been using the old plugin and had no complaints about how it worked.

What with the fact that all of the sudden everything is async because it's "oc" based, and you have to do your own complete status checks because a failed build won't automatically break a stage any more, ... This is just too low level to be useful, because it shifts the burden of knowing the internals of every single operation outcome back to the pipeline instead of keeping them simple and declarative.

I understand the greater flexibility this new plug-in gives to people who want to hack away at their pipelines, but I argue this is a giant step backwards for users that already have their pipelines and are happy with the fact they are simple.

**Comment by mbecca** [comment]
Hi,

"openshift.selector("dc/simple").rollout().latest()" start a new deploy?

You have any trigger in DeploymentConfig? (for example: ImageStreamChange..) 

Thanks in advance!

M.

**Comment by TomSchmitz** [comment]
Ich bin bis 11.02.2019 abwesend.
I am out of the office until 11.02.2019.

Ich werde Ihre Nachricht nach meiner Rückkehr beantworten.

In dringenden Fällen wenden Sie sich bitte an Frank Becker (AC-6606)


Hinweis: Dies ist eine automatische Antwort auf Ihre Nachricht  "Re:
[openshift/jenkins-client-plugin] Any chance of ever having nice, clean,
declarative steps again? (#227)" gesendet am 03.02.2019 00:02:03.

Diese ist die einzige Benachrichtigung, die Sie empfangen werden, während
diese Person abwesend ist.

**Comment by benko** [comment]
Hey, @mbecca,

> "openshift.selector("dc/simple").rollout().latest()" start a new deploy?

Yes. There are actually a number of different ways to start a new deployment using the plugin, this is just one of them.

> You have any trigger in DeploymentConfig? (for example: ImageStreamChange..)

No triggers, of course. With Jenkins, you want Jenkins to drive the pipeline, not triggers.

Think of the following (very common) use-case: developers have no direct access to the cluster, except to the Jenkins console, where the only thing they can do is start a pipeline and view the logs. Their projects are orchestrated, start-to-finish, by Jenkins. Every single step must be accounted for (and logged) in the pipeline.

**Comment by TomSchmitz** [comment]
Ich bin bis 11.02.2019 abwesend.
I am out of the office until 11.02.2019.

Ich werde Ihre Nachricht nach meiner Rückkehr beantworten.

In dringenden Fällen wenden Sie sich bitte an Frank Becker (AC-6606)


Hinweis: Dies ist eine automatische Antwort auf Ihre Nachricht  "Re:
[openshift/jenkins-client-plugin] Any chance of ever having nice, clean,
declarative steps again? (#227)" gesendet am 03.02.2019 07:10:18.

Diese ist die einzige Benachrichtigung, die Sie empfangen werden, während
diese Person abwesend ist.

**Comment by mbecca** [comment]
> 
> Ich bin bis 11.02.2019 abwesend. I am out of the office until 11.02.2019. Ich werde Ihre Nachricht nach meiner Rückkehr beantworten. In dringenden Fällen wenden Sie sich bitte an Frank Becker (AC-6606) Hinweis: Dies ist eine automatische Antwort auf Ihre Nachricht "Re: [openshift/jenkins-client-plugin] Any chance of ever having nice, clean, declarative steps again? (#227)" gesendet am 03.02.2019 00:02:03. Diese ist die einzige Benachrichtigung, die Sie empfangen werden, während diese Person abwesend ist.

Hey, @benko , 

thanks for all the answers, it was very helpful!

Regards, M.

**Comment by TomSchmitz** [comment]
Ich bin bis 11.02.2019 abwesend.
I am out of the office until 11.02.2019.

Ich werde Ihre Nachricht nach meiner Rückkehr beantworten.

In dringenden Fällen wenden Sie sich bitte an Frank Becker (AC-6606)


Hinweis: Dies ist eine automatische Antwort auf Ihre Nachricht  "Re:
[openshift/jenkins-client-plugin] Any chance of ever having nice, clean,
declarative steps again? (#227)" gesendet am 05.02.2019 23:59:47.

Diese ist die einzige Benachrichtigung, die Sie empfangen werden, während
diese Person abwesend ist.

**Comment by gabemontero** [comment]
Hi @benko 

Sidebar: as creator of the old plugin, I could not help but smile at your reaction/opinions :-) especially in light of having to maintain this new plugin

Now a core dump of the particulars:
- generally speaking, this plugin has received more positive feedback .... folks like the jenkins/pipelines/groovy syntax sugar
- the old plugin was too expensive to maintain / evolve because of its native java REST client, which has been deprecated; this plugin's use of the `oc` binary made it much easier to meet the demand for features and picking up of changes in `oc` that users were always asking about
- don't forget, sh 'oc start build simple -f' is still possible ... in many cases, sh 'oc ...' was on par with what the old plugin provided, and you can avoid the syntax sugar  ... we are in more and more cases recommending such simple approaches when questions come up

bottom line: 
- old plugin isn't coming back, 
- major changes to this plugin wrt syntax are not planned currently ... if there was something to motivate major changes, it would be the restrictions with the declarative pipeline syntax and the pipeline global variable plug point within Jenkins that his plugin leverages
- we are monitoring knative / jenkins X / other CI/CI alternatives underneath knative, both within openshift and in the broader community, and we'll see if/how this plugin plays with all that when things mature

Probably not what you wanted to hear entirely, but hopefully the details provide some context at least.
